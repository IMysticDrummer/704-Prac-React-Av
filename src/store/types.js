export const AUTH_LOGIN_REQUEST = 'AUTH_LOGIN_REQUEST';
export const AUTH_LOGIN_SUCCESS = 'AUTH_LOGIN_SUCCESS';
export const AUTH_LOGIN_FAILURE = 'AUTH_LOGIN_FAILURE';
export const AUTH_LOGOUT = 'AUTH_LOGOUT';
export const TAGS_REQUEST = 'TAGS_REQUEST';
export const TAGS_SUCCESS = 'TAGS_SUCCESS';
export const TAGS_FAILURE = 'TAGS_FAILURE';
export const UI_RESET_ERROR = 'UI_RESET_ERROR';
