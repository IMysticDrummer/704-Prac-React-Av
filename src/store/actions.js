//import { getTags } from '../components/AdvertsPage/service.js';
import {
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGIN_FAILURE,
  TAGS_REQUEST,
  TAGS_SUCCESS,
  TAGS_FAILURE,
  UI_RESET_ERROR,
  AUTH_LOGOUT_REQUEST,
  AUTH_LOGOUT_SUCCESS,
  AUTH_LOGOUT_CANCEL,
} from './types.js';

export function authLoginRequest() {
  return {
    type: AUTH_LOGIN_REQUEST,
  };
}

export function authLoginSuccess() {
  return {
    type: AUTH_LOGIN_SUCCESS,
  };
}

export function authLoginFailure(error) {
  return {
    type: AUTH_LOGIN_FAILURE,
    payload: error,
    error: true,
  };
}

export const authLogin = (credentials) => {
  const { email, password, remember } = credentials;
  return async function (dispatch, getState, { api, router }) {
    try {
      dispatch(authLoginRequest());
      await api.auth.login({ email, password }, remember);
      dispatch(authLoginSuccess());
      const to = router.state.location.state?.from?.pathname || '/';
      router.navigate(to, { replace: true });
    } catch (error) {
      dispatch(authLoginFailure(error));
    }
  };
};

export function authLogoutRequest() {
  return {
    type: AUTH_LOGOUT_REQUEST,
  };
}
export function authLogoutSuccess() {
  return {
    type: AUTH_LOGOUT_SUCCESS,
  };
}
export function authLogoutCancel() {
  return {
    type: AUTH_LOGOUT_CANCEL,
  };
}

export function authLogout(response) {
  return async function (dispatch, getState, { api, router }) {
    if (response) {
      try {
        await api.auth.handleLogout();
        dispatch(authLogoutSuccess());
        const to = router.state.location.state?.from?.pathname || '/';
        router.navigate(to, { replace: true });
      } catch (error) {
        //TODO tratar el error
      }
    } else {
      dispatch(authLogoutCancel());
    }
  };
}

export function tagsRequest() {
  return {
    type: TAGS_REQUEST,
  };
}

export function tagsSuccess(tags) {
  return {
    type: TAGS_SUCCESS,
    payload: tags,
  };
}
export function tagsFailure(error) {
  return {
    type: TAGS_FAILURE,
    payload: error,
    error: true,
  };
}

export function getTagsAction() {
  return async function (dispatch, getState) {
    try {
      dispatch(tagsRequest());
      //const tags = await getTags();
      //dispatch(tagsSuccess(tags));
    } catch (error) {
      dispatch(tagsFailure());
      throw error;
    }
  };
}

export function uiResetError() {
  return {
    type: UI_RESET_ERROR,
  };
}
