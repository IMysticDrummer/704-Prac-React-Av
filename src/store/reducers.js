import {
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGIN_FAILURE,
  TAGS_LOAD_SUCCESS,
  UI_RESET_ERROR,
  AUTH_LOGOUT_SUCCESS,
} from './types.js';

export const defaultState = {
  auth: false,
  tags: {
    areLoaded: false,
    data: [],
  },
  ui: {
    isLoading: false,
    error: null,
  },
};

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_REQUEST:
      return false;
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGIN_FAILURE:
      return false;
    case AUTH_LOGOUT_SUCCESS:
      return false;
    default:
      return state;
  }
}

export function tags(state = defaultState.tags, action) {
  switch (action.type) {
    case TAGS_LOAD_SUCCESS:
      return { areLoaded: true, data: action.payload };
    default:
      return state;
  }
}

export function ui(state = defaultState.ui, action) {
  // Any case in error. For example AUTH_LOGIN_FAILURE
  if (action.error) {
    return { error: action.payload, isLoading: false };
  }

  if (/_REQUEST$/.test(action.type)) {
    return {
      error: null,
      isLoading: true,
    };
  }
  //Toda acci√≥n que acabe en SUCCESS con expresiones regulares
  if (/_SUCCESS$/.test(action.type)) {
    return {
      error: null,
      isLoading: false,
    };
  }
  if (/_CANCEL$/.test(action.type)) {
    return {
      ...state,
      isLoading: false,
      error: null,
    };
  }

  if (action.type === UI_RESET_ERROR) {
    return {
      ...state,
      error: null,
    };
  }

  return state;
}
