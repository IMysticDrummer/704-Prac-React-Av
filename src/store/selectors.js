export const getIsLogged = (state) => state.auth.state;

export const getIsLogginout = (state) => state.auth.askingLogout;

export const getUi = (state) => state.ui;

export const areTagsLoaded = (state) => state.tags.areLoaded;

export const getTags = (state) => (state.tags.areLoaded ? state.tags.data : []);

export const areAdsLoaded = (state) => state.ads.areLoaded;

export const getAds = (state) => (state.ads.areLoaded ? state.ads.data : []);
