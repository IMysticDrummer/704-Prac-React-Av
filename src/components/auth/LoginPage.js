import { useState } from 'react';
import EnterElement from '../common/EnterElement';
import classNames from 'classnames';
import styles from './LoginPage.module.css';

import ErrorElement from '../common/ErrorElement';
import { Button } from '../common/Button';
import styled from 'styled-components';
import Spinner from '../common/Spinner';
import { useDispatch, useSelector } from 'react-redux';
import { authLogin, uiResetError } from '../../store/actions';
import { getUi } from '../../store/selectors';

export const LoginPage = ({ className, ...props }) => {
  const [password, setPassword] = useState([]);
  const [email, setEmail] = useState([]);
  const [remember, setRemember] = useState(false);
  const dispatch = useDispatch();
  const { isLoading: isFetching, error } = useSelector(getUi);

  const enterElementHandleChange = (event) => {
    if (event.target.name === 'password') {
      setPassword(event.target.value);
    }
    if (event.target.name === 'email') {
      setEmail(event.target.value);
    }
    if (event.target.name === 'remember') {
      setRemember((remember) => !remember);
    }
  };

  const disableButton = () => {
    return !(email.length !== 0 && password.length !== 0 && !isFetching);
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    dispatch(authLogin({ email, password, remember }));
  };

  const handleErrorMessageClick = (event) => {
    event.preventDefault();
    dispatch(uiResetError());
  };

  return (
    <section className={classNames(styles.LoginPage, className)}>
      <h2>Login</h2>
      <form onSubmit={handleSubmit}>
        {isFetching && <Spinner />}
        <EnterElement
          labelText='Enter your email'
          type='email'
          name='email'
          placeholder='someone@mail.com'
          onChange={enterElementHandleChange}
          value={email}
        />
        <EnterElement
          labelText='Enter your password'
          type='password'
          name='password'
          onChange={enterElementHandleChange}
          value={password}
        />
        <EnterElement
          labelText='Remember me'
          type='checkbox'
          name='remember'
          onChange={enterElementHandleChange}
          checked={remember}
        />

        <Button
          variant='primary'
          type='submit'
          disabled={disableButton()}
        >
          Login
        </Button>
      </form>
      {error?.status === 500 && (
        <ErrorElement
          error={error}
          altMessage='Test your cable'
          handleErrorMessageClick={handleErrorMessageClick}
        />
      )}
      {error?.status === 401 && (
        <ErrorElement
          error={error}
          altMessage='Unathorized'
          handleErrorMessageClick={handleErrorMessageClick}
        />
      )}
    </section>
  );
};

const StyledLoginPage = styled(LoginPage)`
  background-color: ${'var(--main-bg-color)' || 'lightblue'};
  color: ${'var(--main-color)' || 'darkblue'};
`;

export default StyledLoginPage;
