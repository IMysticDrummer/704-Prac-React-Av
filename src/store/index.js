import { combineReducers, createStore, applyMiddleware } from 'redux';
import * as reducers from './reducers';
import { composeWithDevTools } from '@redux-devtools/extension';
import thunk from 'redux-thunk';
import * as auth from '../components/auth/service';

const reducer = combineReducers(reducers);
const middlewares = [thunk.withExtraArgument({ api: { auth } })];

export default function configureStore(preloadState) {
  const store = createStore(
    reducer,
    preloadState,
    composeWithDevTools(applyMiddleware(...middlewares))
  );
  return store;
}
