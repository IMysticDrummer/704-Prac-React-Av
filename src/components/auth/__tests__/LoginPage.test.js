import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { authLogin, uiResetError } from '../../../store/actions';
import { defaultState } from '../../../store/reducers';
import { EnhancedLoginPage } from '../LoginPage';

jest.mock('../../../store/actions');

describe('Test de la pÃ¡gina LoginPage', () => {
  const store = {
    getState: () => defaultState,
    dispatch: () => {},
    subscribe: () => {},
  };
  const errorState401 = {
    ...defaultState,
    ui: { isLoading: false, error: { status: 401 } },
  };
  const errorState500 = {
    ...defaultState,
    ui: { isLoading: false, error: { status: 500 } },
  };
  const storeWithError401 = { ...store, getState: () => errorState401 };
  const storeWithError500 = { ...store, getState: () => errorState500 };

  const renderComponent = (store) =>
    render(
      <Provider store={store}>
        <EnhancedLoginPage />
      </Provider>
    );

  const validCredentials = {
    email: 'prueba@yo.com',
    password: '123456',
    remember: 'on',
  };

  test('Snapshot', () => {
    const { container } = renderComponent(store);
    expect.assertions(1);
    expect(container).toMatchSnapshot();
  });

  test('submit button must be disabled if email field or password field are empty', () => {
    renderComponent(store);
    const emailInput = screen.getByLabelText(/email/);
    const passwordInput = screen.getByLabelText(/password/);
    const submitButton = screen.getByRole('button');

    expect.assertions(4);
    expect(submitButton).toBeDisabled();
    fireEvent.change(emailInput, { target: { value: validCredentials.email } });
    expect(submitButton).toBeDisabled();
    fireEvent.change(passwordInput, {
      target: { value: validCredentials.password },
    });
    fireEvent.change(emailInput, { target: { value: null } });
    expect(submitButton).toBeDisabled();
    fireEvent.change(emailInput, { target: { value: validCredentials.email } });
    expect(submitButton).toBeEnabled();
  });

  it('shoud dispatch authLogin action when username and password exists and click "Login"', () => {
    renderComponent(store);

    const emailInput = screen.getByLabelText(/email/);
    const passwordInput = screen.getByLabelText(/password/);
    const rememberInput = screen.getByLabelText(/Remember/);
    const submitButton = screen.getByRole('button');

    fireEvent.change(emailInput, { target: { value: validCredentials.email } });
    fireEvent.change(passwordInput, {
      target: { value: validCredentials.password },
    });
    fireEvent.click(rememberInput);

    fireEvent.click(submitButton);
    expect.assertions(1);
    expect(authLogin).toHaveBeenCalledWith(validCredentials);
  });

  describe('Error messages in LoginPage', () => {
    it("shouldn't appear error if there's no an error", async () => {
      renderComponent(store);
      expect.assertions(2);
      let message;
      try {
        const message = await screen.findByText(/401/);
      } catch {
        expect(message).not.toBeDefined();
      }
      try {
        message = await screen.findByText(/500/);
      } catch {
        expect(message).not.toBeDefined();
      }
    });
    it("should appear Unathorized-401 if there's an 401 error", async () => {
      renderComponent(storeWithError401);
      const message = await screen.findByText(/401/);
      expect.assertions(1);
      expect(message).toBeVisible();
    });
    it("should appear 500 error if there's an 500 error", async () => {
      renderComponent(storeWithError500);
      const message = await screen.getByText(/500/);
      expect.assertions(1);
      expect(message).toBeVisible();
    });
    it('should disappear when click over it', () => {
      renderComponent(storeWithError401);
      const message = screen.getByText(/continue/);
      expect.assertions(2);
      expect(message).toBeDefined();
      fireEvent.click(message);
      expect(uiResetError).toHaveBeenCalled();
    });
  });
});
