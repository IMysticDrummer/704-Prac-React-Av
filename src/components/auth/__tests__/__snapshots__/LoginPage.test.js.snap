// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test de la p√°gina LoginPage Snapshot 1`] = `
<div>
  <section
    class="LoginPage"
  >
    <h2>
      Login
    </h2>
    <form>
      <article
        class="enterElementClass sc-bcXHqe hAGHhm"
      >
        <label>
          <span>
            Enter your email
          </span>
          <input
            name="email"
            placeholder="someone@mail.com"
            type="email"
            value=""
          />
        </label>
      </article>
      <article
        class="enterElementClass sc-bcXHqe hAGHhm"
      >
        <label>
          <span>
            Enter your password
          </span>
          <input
            name="password"
            type="password"
            value=""
          />
        </label>
      </article>
      <article
        class="enterElementClass sc-bcXHqe hAGHhm"
      >
        <label>
          <span>
            Remember me
          </span>
          <input
            name="remember"
            type="checkbox"
          />
        </label>
      </article>
      <button
        class="sc-gswNZR fjeFES"
        disabled=""
        type="submit"
      >
        Login
      </button>
    </form>
  </section>
</div>
`;
